src/machine/machine_rp2040_pio.go

```diff
@@ -0,0 +1,397 @@
+// +build rp2040
+
+package machine
+
+import (
+	"device/rp"
+	"runtime/volatile"
+	"unsafe"
+)
+
+var (
+	PIO0 = &PIO{
+		Device: rp.PIO0,
+	}
+	PIO1 = &PIO{
+		Device: rp.PIO1,
+	}
+)
+
+const (
+	REG_ALIAS_RW_BITS  = 0x0 << 12
+	REG_ALIAS_XOR_BITS = 0x1 << 12
+	REG_ALIAS_SET_BITS = 0x2 << 12
+	REG_ALIAS_CLR_BITS = 0x3 << 12
+)
+
+// PIO represents one of the two PIO peripherals in the RP2040
+type PIO struct {
+	// Bitmask of used instruction space
+	usedSpaceMask uint32
+
+	// Index of this PIO (0 or 1)
+	Index uint8
+
+	// StateMachines provides access to the 4 state machines in this PIO
+	StateMachines [4]PIOStateMachine
+
+	// Device is the actual hardware device
+	Device *rp.PIO0_Type
+}
+
+type PIOStateMachineReg uint8
+
+const (
+	PIOStateMachineClkDivReg    PIOStateMachineReg = 0
+	PIOStateMachineExecCtrlReg                     = 4
+	PIOStateMachineShiftCtrlReg                    = 8
+	PIOStateMachineAddrReg                         = 12
+	PIOStateMachineInstrReg                        = 16
+	PIOStateMachinePinCtrlReg                      = 20
+)
+
+// PIOStateMachine represents one of the four state machines in a PIO
+type PIOStateMachine struct {
+	// The PIO containing this state machine
+	PIO *PIO
+
+	// Index of this state machine
+	Index uint8
+}
+
+// PIOStateMachineConfig holds the configuration for a PIO state
+// machine.
+//
+// This type is used by code generated by pioasm, in the RP2040
+// c-sdk - any changes should be backwards compatible.
+type PIOStateMachineConfig struct {
+	ClkDiv    uint32
+	ExecCtrl  uint32
+	ShiftCtrl uint32
+	PinCtrl   uint32
+}
+
+// PIOProgram holds the assembled PIO code
+//
+// This type is used by code generated by pioasm, in the RP2040
+// c-sdk - any changes should be backwards compatible.
+type PIOProgram struct {
+	// Instructions holds the binary code in 16-bit words
+	Instructions []uint16
+
+	// Origin indicates where in the PIO execution memory
+	// the program must be loaded, or -1 if the code is
+	// position independant
+	Origin int8
+}
+
+func (pio *PIO) Configure() {
+	pio.Index = 0
+	if pio == PIO1 {
+		pio.Index = 1
+	}
+
+	for i := 0; i < 4; i++ {
+		pio.StateMachines[i].PIO = pio
+		pio.StateMachines[i].Index = uint8(i)
+	}
+}
+
+// AddProgram loads a PIO program into PIO memory
+func (pio *PIO) AddProgram(program *PIOProgram) uint8 {
+	offset := pio.findOffsetForProgram(program)
+	if offset < 0 {
+		panic("no program space")
+	}
```

Comment:
I think it would be better to return an error here instead of panicking.

Reply:
Yeah.  Good call.  I'll update.  I copied it from the c-sdk, where they do this:
```c
// these assert if unable
uint pio_add_program(PIO pio, const pio_program_t *program) {
    uint32_t save = hw_claim_lock();
    int offset = _pio_find_offset_for_program(pio, program);
    if (offset < 0) {
        panic("No program space");
    }
    _pio_add_program_at_offset(pio, program, (uint)offset);
    hw_claim_unlock(save);
    return (uint)offset;
}
```

Reply:
FYI, they also have a way to check before calling pio_add_program:
```c
bool pio_can_add_program(PIO pio, const pio_program_t *program) {
    uint32_t save = hw_claim_lock();
    bool rc =  -1 != _pio_find_offset_for_program(pio, program);
    hw_claim_unlock(save);
    return rc;
}
```
In Go, we either need to return an error, or have a function that panics and a `TryXXXX` variant.